### ISSUE TITLE
Declare `@Synchronized` lock fields as `Object[]` instead of `Object`

### ISSUE THREAD

**Issue by prdoyle** [rzwitserloot/lombok#2604]
**Describe the feature**
Instead of declaring the lock fields (`$lock`, `$LOCK`, and `readLock`) as `Object`, they should be declared as `Object[]`.

Declaring a lock object field as type `Object` is a uniquely poor choice, because it is the one and only datatype in Java that could be either an array or a non-array object. Implementations such as OpenJ9 use a different offset for the lock word depending on whether or not the object is an array. That means if (and only if) your field is declared as an `Object`, the jit-generated code must contain a branching sequence that tests whether the object is an array.

For any other field type, this extra code is not needed.

**Describe the target audience**
Users of `@Synchronized` on OpenJ9 and any other JVM using different lock word offsets for arrays versus non-arrays.

**Comment by Rawi01** [comment]
I never thought about something like this and I only found [this](https://blog.openj9.org/2019/04/02/lock-nursery/). I might have missed something but I think that OpenJ9 uses lock words for `j.l.Object` and completly excludes arrays. If I assume that the lock word stuff is faster than regular locks an array implementation will be slower than the current one.
Can you provide more details why you think that an array is faster?

**Comment by rspilker** [comment]
I agree with @Rawi01 that possibly a better fix would be to replace the Object[] with something else that is an Object, Serializable and small. Something like `new Integer(0)`.
